hail/src/main/scala/is/hail/types/physical/PCanonicalNDArray.scala

```diff
@@ -151,16 +146,30 @@ final case class PCanonicalNDArray(elementType: PType, nDims: Int, required: Boo
           codeSoFar,
           bytesAway := bytesAway + requestedIndex * stridesTuple(strideIndex))
       },
-      bytesAway + data.pType.elementOffset(dataStore, data.pType.loadLength(dataStore), 0)
+      bytesAway + data.pType.firstElementOffset(dataStore, data.pType.loadLength(dataStore))
     ))
   }
 
-  def setElement(indices: IndexedSeq[Value[Long]], ndAddress: Value[Long], newElement: Code[_], mb: EmitMethodBuilder[_]): Code[Unit] = {
-    Region.storeIRIntermediate(this.elementType)(getElementAddress(indices, ndAddress, mb), newElement)
+  private def getElementAddressFromDataPointerAndStrides(indices: IndexedSeq[Value[Long]], dataFirstElementPointer: Value[Long], strides: IndexedSeq[Value[Long]], cb: EmitCodeBuilder): Code[Long] = {
+    val bytesAway = cb.newLocal[Long]("nd_get_element_address_bytes_away")
+
+    coerce[Long](Code(
```

Comment:
append this stuff to the CB instead of using the `Code(...)` constructor?

Reply:
This was a lot less trivial to fix then I thought. There were several places calling `loadElement` (which calls this method as a helper) in a way that assumed it was returning a lazily evaluated code and wouldn't have any side effects. The changes to `NDArrayMatMul` in the emitter and to `EColumnMajorNDArray` both were necessary to avoid this.

Reply:
oof, yup, that's a risk.

Reply:
A pattern we used a lot was emitting a lazily-inserted Code[T] at the start of a block, when we could have generated it in a loop or something. This is painful to fix.